envoy/stats/allocator.h

```diff
@@ -32,20 +32,23 @@ class Allocator {
    * @param name the full name of the stat.
    * @param tag_extracted_name the name of the stat with tag-values stripped out.
    * @param tags the tag values.
+   * @param is_custom_metric true if this is a custom metric, false otherwise.
    * @return CounterSharedPtr a counter.
    */
   virtual CounterSharedPtr makeCounter(StatName name, StatName tag_extracted_name,
-                                       const StatNameTagVector& stat_name_tags) PURE;
+                                       const StatNameTagVector& stat_name_tags,
```

Comment:
All these changes make me think that we've now reached a threshold where it makes sense to have a common StatsOptions struct:

```
struct StatsOptions {
  StatsOptions(StatName) : name_(name) {}
  StatName name_;
  const StatNameTagVector stat_name_tags_;
  bool is_custom_metric_ = false;
};
```
those would be common to all the stat types; you could still pass import-mode and units separately to gauges and histograms.

This way if another initialization parameter is added later it can be added to the structure rather than changing 100s of overrides and call-sites.

If you switch to that in this PR, this PR will not grow bigger than it already is. But you'll make it a lot easier to make these sorts of changes later. You'll also inherently fix a bug I found in your new code below where you were not transmitting 'is_custom_metric' properly in a delegate.

Reply:
slight variant on this: have non-virtual derived types CounterOptions, GaugeOptions, HistogramOptions, that add the specific fields for those. Name the base class BaseOptions (it's already in namespace `Stats`).

Reply:
Sounds good, but I wonder how do you expect these options to be used in gaugeFromStatNameWithTags or gaugeFromStatName for example?
StatNameTagVector  in the "BaseOption" is initially be extracted from StatNameTagVectorOptConstRef inside of its implementation. So we won't be able to pass these new option struct directly to *FromStatNameWithTags implementation, and therefore we still end up adding a new option to every overrides just like in this PR. thoughts?

Reply:
I think we could drop the WithStatName variants with this change, which would reduce the large surface-area of this API. It would change all existing call. I guess in that case the base struct should hold a const ref:
```
struct StatsOptions {
  StatsOptions(StatName) : name_(name) {}
  StatName name_;
  StatNameTagVectorOptConstRef stat_name_tags_;
  bool is_custom_metric_ = false;
};
```

I think maybe I'll retract this suggestion for now, for this PR, as it probably would in fact create more churn.